it/src/main/java/org/corfudb/universe/universe/vm/VmUniverse.java

```diff
@@ -0,0 +1,319 @@
+
+        Map<String, VirtualMachine> deployment = new HashMap<>();
+
+        List<CompletableFuture<VirtualMachine>> asyncDeployment = universeParams.getVmIpAddresses()
+                .keySet()
+                .stream()
+                .map(this::deployVmAsync)
+                .collect(Collectors.toList());
+
+        asyncDeployment
+                .stream()
+                .map(CompletableFuture::join)
+                .forEach(vm -> {
+                    universeParams.updateIpAddress(vm.getName(), vm.getGuest().getIpAddress());
+                    deployment.put(vm.getName(), vm);
+                });
+
+        log.info("The deployed VMs are: {}", universeParams.getVmIpAddresses());
+
+        vms.putAll(deployment);
+    }
+
+    /**
+     * Deploy a {@link Group} on existing VMs according to input parameter.
+     */
+    @Override
+    protected Group deployGroup(GroupParams groupParams) {
+        switch (groupParams.getNodeType()) {
+            case CORFU_SERVER:
+                VmCorfuCluster cluster = VmCorfuCluster.builder()
+                        .universeParams(universeParams)
+                        .params(ClassUtils.cast(groupParams))
+                        .vms(ImmutableMap.copyOf(vms))
+                        .build();
+
+                return cluster.deploy();
+            case CORFU_CLIENT:
+                throw new UniverseException("Not implemented corfu client. Group config: " + groupParams);
+            default:
+                throw new UniverseException("Unknown node type");
+        }
+    }
+
+    /**
+     * Shutdown the {@link Universe} by stopping each of its {@link Group}.
+     */
+    @Override
+    public void shutdown() {
+        if (destroyed.getAndSet(true)) {
+            log.info("Can't shutdown vm universe. Already destroyed");
+            return;
+        }
+
+        log.info("Shutdown the universe: {}, params: {}", universeId, groups);
+        shutdownGroups();
+    }
+
+    @Override
+    public Universe add(GroupParams groupParams) {
+        throw new UnsupportedOperationException("Not implemented");
+    }
+
+    private CompletableFuture<VirtualMachine> deployVmAsync(String vmName) throws NodeException {
+        log.info("Deploy vm asynchronously: {}", vmName);
+        return CompletableFuture.supplyAsync(() -> deployVm(vmName), executor);
+    }
+
+    @Override
+    protected void shutdownGroups() {
+        super.shutdownGroups();
+
+        ImmutableSet<String> groupNames = universeParams.getGroups().keySet();
+
+        groupNames.forEach(groupName -> {
+            Group group = groups.get(groupName);
+            if (group != null) {
+                return;
+            }
+            universeParams
+                    .getGroupParams(groupName, GroupParams.class)
+                    .<VmServerParams>getNodesParams()
+                    .forEach(serverParams -> {
+                        CorfuServerOnVm node = CorfuServerOnVm.builder()
```

Comment:
This looks a little off. In order to destroy a node, we need to create it?

Reply:
Fixed, don't need to create nodes anymore